Test dependency on `generics-sop` needs upper bound (or arbitrary instances need updating)

The package states that it depends on `generics-sop >= 0.2`, but attempts to compile with `generics-sop-0.4.0.1` results in:

```
Building test suite 'test' for uri-bytestring-0.3.2.1..
[1 of 4] Compiling URI.ByteString.Arbitrary ( test/URI/ByteString/Arbitrary.hs, dist/build/test/test-tmp/URI/ByteString/Arbitrary.o )

test/URI/ByteString/Arbitrary.hs:42:28: error:
    • Could not deduce (SOP.All SOP.SListI xs)
        arising from a use of ‘SOP.hsequence’
      from the context: (SOP.SListI xs, SOP.AllF (SOP.All Arbitrary) xs,
                         SOP.AllF SOP.SListI xs)
        bound by the type signature for:
                   sopArbitrary' :: forall (xs :: [[*]]).
                                    (SOP.SListI xs, SOP.AllF (SOP.All Arbitrary) xs,
                                     SOP.AllF SOP.SListI xs) =>
                                    Gen (SOP.SOP SOP.I xs)
        at test/URI/ByteString/Arbitrary.hs:(39,1)-(41,27)
    • In the first argument of ‘map’, namely ‘SOP.hsequence’
      In the expression: map SOP.hsequence
      In the first argument of ‘oneof’, namely
        ‘(map SOP.hsequence $ SOP.apInjs_POP $ SOP.hcpure p arbitrary)’
   |
42 | sopArbitrary' = oneof (map SOP.hsequence $ SOP.apInjs_POP $ SOP.hcpure p arbitrary)
   |                            ^^^^^^^^^^^^^

test/URI/ByteString/Arbitrary.hs:42:61: error:
    • Could not deduce (SOP.All (SOP.All Arbitrary) xs)
        arising from a use of ‘SOP.hcpure’
      from the context: (SOP.SListI xs, SOP.AllF (SOP.All Arbitrary) xs,
                         SOP.AllF SOP.SListI xs)
        bound by the type signature for:
                   sopArbitrary' :: forall (xs :: [[*]]).
                                    (SOP.SListI xs, SOP.AllF (SOP.All Arbitrary) xs,
                                     SOP.AllF SOP.SListI xs) =>
                                    Gen (SOP.SOP SOP.I xs)
        at test/URI/ByteString/Arbitrary.hs:(39,1)-(41,27)
    • In the second argument of ‘($)’, namely ‘SOP.hcpure p arbitrary’
      In the second argument of ‘($)’, namely
        ‘SOP.apInjs_POP $ SOP.hcpure p arbitrary’
      In the first argument of ‘oneof’, namely
        ‘(map SOP.hsequence $ SOP.apInjs_POP $ SOP.hcpure p arbitrary)’
   |
42 | sopArbitrary' = oneof (map SOP.hsequence $ SOP.apInjs_POP $ SOP.hcpure p arbitrary)
   |                                                             ^^^^^^^^^^^^^^^^^^^^^^
[2 of 4] Compiling URI.ByteStringQQTests ( test/URI/ByteStringQQTests.hs, dist/build/test/test-tmp/URI/ByteStringQQTests.o )
```

--- a/test/URI/ByteString/Arbitrary.hs
+++ b/test/URI/ByteString/Arbitrary.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP                   #-}
 {-# LANGUAGE ConstraintKinds       #-}
 {-# LANGUAGE DataKinds             #-}
 {-# LANGUAGE FlexibleContexts      #-}
@@ -11,7 +12,10 @@ module URI.ByteString.Arbitrary where
 import           Control.Applicative
 import           Data.Proxy                (Proxy (..))
 import qualified Generics.SOP              as SOP
+#if MIN_VERSION_generics_sop(0,4,0)
+#else
 import qualified Generics.SOP.Constraint   as SOP
+#endif
 import qualified Generics.SOP.GGP          as SOP
 import           GHC.Generics              (Generic)
 import           Test.QuickCheck
@@ -29,15 +33,28 @@ sopArbitrary
   :: ( SOP.SListI (SOP.GCode b)
      , Generic b
      , SOP.GTo b
+#if MIN_VERSION_generics_sop(0,4,0)
+     , SOP.All SOP.SListI (SOP.GCode b)
+     , SOP.All (SOP.All Arbitrary) (SOP.GCode b)
+#else
      , SOP.AllF SOP.SListI (SOP.GCode b)
      , SOP.AllF (SOP.All Arbitrary) (SOP.GCode b)
+#endif
      )
   => Gen b
 sopArbitrary = fmap SOP.gto sopArbitrary'
 
 
 sopArbitrary'
-  :: (SOP.SListI xs, SOP.AllF (SOP.All Arbitrary) xs, SOP.AllF SOP.SListI xs)
+  :: ( SOP.SListI xs
+#if MIN_VERSION_generics_sop(0,4,0)
+     , SOP.All SOP.SListI xs
+     , SOP.All (SOP.All Arbitrary) xs
+#else
+     , SOP.AllF SOP.SListI xs
+     , SOP.AllF (SOP.All Arbitrary) xs
+#endif
+     )
   => Gen (SOP.SOP SOP.I xs)
 sopArbitrary' = oneof (map SOP.hsequence $ SOP.apInjs_POP $ SOP.hcpure p arbitrary)
   where

--- a/uri-bytestring.cabal
+++ b/uri-bytestring.cabal
@@ -89,7 +89,7 @@ test-suite test
     , semigroups
     , transformers
     , containers
-    , generics-sop >= 0.3 && < 0.5
+    , generics-sop >= 0.2 && < 0.5
   default-language:    Haskell2010
 
   if flag(lib-Werror)

--- a/uri-bytestring.cabal
+++ b/uri-bytestring.cabal
@@ -1,5 +1,5 @@
 name:                uri-bytestring
-version:             0.3.2.1
+version:             0.3.2.2
 synopsis:            Haskell URI parsing as ByteStrings
 description: uri-bytestring aims to be an RFC3986 compliant URI parser that uses efficient ByteStrings for parsing and representing the URI data.
 license:             BSD3
@@ -89,7 +89,7 @@ test-suite test
     , semigroups
     , transformers
     , containers
-    , generics-sop >= 0.2
+    , generics-sop >= 0.3 && < 0.5
   default-language:    Haskell2010
 
   if flag(lib-Werror)

